lib/os/CMakeLists.txt

```diff
@@ -24,6 +24,7 @@ zephyr_sources(
   heap-validate.c
   bitarray.c
   multi_heap.c
+  winstream.c
```

Comment:
Let's add a Kconfig option for this so we don't slow down the build for all that don't use this.

Reply:
Easy enough.  But FWIW: I don't know what I buy that analysis.  C file builds are wildly parallel on almost all the environments we care about.    Extra CMake logic ties up a single core before any compiler can be launched at all, and it's already dominating the build.   A quick test on my box:

```
$ time cmake -B build -GNinja -DBOARD=qemu_x86_64 samples/hello_world
...
real	0m1.474s
user	0m1.286s
sys	0m0.173s

$ time (cd build; ninja)
...
real	0m0.799s
user	0m2.516s
sys	0m0.955s
```

Again, not like it matters here.  But IMHO we need to be spending our development optimization effort trying to reduce the amount of cmake usage, not increase it.

Reply:
A big problem with additional Kconfigs is combinatorial explosion. I mean many Kconfig combinations are never tested, it's frequent some combinations don't even compile! kconfig.c has a very useful "randconfig" option but the orphaned kconfig.py used by Zephyr has not.

To speed up compilation from scratch I think cleaning up #includes (#41543) seems more effort but also much more promising because affecting many .c files at once. It also important "modularity hygiene".

The performance of incremental C compilation is not a problem because it's... incremental and highly parallel. As mentioned by @andyross it's entirely dominated by CMake itself and/or the various linking scripts and post-processing.

I think more Kconfig matters but for a very different optimization: space. In SOF we very very often run out of space on some platforms. Would an additional Kconfig here make a difference?

Reply:
Regarding space: not really.  With only a tiny handful of edge cases ("stuff with SYS_INIT() initialization calls doesn't garbage collect" being the big one), everything we link uses -fdata/function-sections and gets included in the final binary only if actually referenced.

Reply:
In other words, turning off tracing with already existing Kconfigs is enough for this code to be removed at link time? Just making sure.

Reply:
Yes, though (getting into DSP specifics) that only gets rid of the code.  The 8k for the trace buffer in window 3 is in the "reserved" block in the bottom of memory that can't be used architecturally for anything else (SOF puts its own stuff in that region too and all the addresses are hard-coded).  I'm hoping to get that fixed in the medium-term future so those regions can be seen by the linker and resized/eliminated/moved, there were some discussions about this in @lyakh 's PR to the internal Zephyr tree.

Reply:
> Yes, though (getting into DSP specifics) that only gets rid of the code. The 8k for the trace buffer in window 3 is in the "reserved" block in the bottom of memory that can't be used architecturally for anything else

So this means right now it's reserved with or without this PR, the latter does not make any difference?

Not familiar with link-time optimizations and not sure about your longer term plans for this but in general maybe any tracing buffer common to various tracing features could be under the control of a new but _invisible_ kconfig symbol which does not contribute to combinatorial explosion because it's not a free variable.
https://docs.zephyrproject.org/2.4.0/guides/kconfig/setting.html#visible-and-invisible-kconfig-symbols 
>  Invisible symbols are not shown in the interactive configuration interfaces, and users have no direct control over their value. They instead get their value from defaults or from other symbols.

Reply:
I'm OK with a promptless AKA invisible Kconfig symbol.

I don't agree with your analysis of the build-time @andyross .

I measure the gcc invocation to be 33ms. Divided by 8 cores we get 4ms.
Note that adding 2 functions to an existing C file would be OK, as it is the invocation
itself that is costly.

We can do a conservative estimate of the Kconfig+CMake runtime.

CMake is practically 0. if(CONFIG_FOO) should take a handful of nanoseconds.

There are 1600 Kconfig source files. If adding one Kconfig symbol
slowed down the build by 4ms, then processing 1600 Kconfig files
(assuming a conservative average of one symbol per source file)
should take 1600 * 4 = 6.4s, which it obviously doesn't.

So the overhead of adding a symbol in Kconfig must be much lower than
4ms, maybe 0.1ms.

In addition to the slowdown there is also noise in the build logs to consider. Many find it useful
to examine the build logs to see what is included in the build.

4ms may not seem like a lot, but OKing this as a policy will make this number grow significantly.

Build systems can be slow with no config required (just include everything and let dynamic dependency tracking sort it out), or fast with explicit configuration. I'm sure you have worked on a slow build system and can appreciate
the build being fast.